### PR TITLE
UCP: add UCP_NUM_EPS option

### PR DESCRIPTION
**Pull request by alex-mikheev**
The option can be used by application to set an estimated number of
endpoints and override number of eps passed in ucp_init()
@yosefe

### CODE REVIEW

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -170,6 +170,14 @@ static ucs_config_field_t ucp_config_table[] = {
    "Also the value has to be bigger than UCX_TM_THRESH to take an effect." ,
    ucs_offsetof(ucp_config_t, ctx.tm_max_bcopy), UCS_CONFIG_TYPE_MEMUNITS},
 
+  {"NUM_EPS", "auto",
+   "An optimization hint of how many endpoints would be created on this context.\n"
+   "Does not affect semantics, but only transport selection criteria and the\n"
+   "resulting performance.\n"
+   " If set to a value different from \"auto\" it will override the value passed\n"
+   "to the ucp_init()",
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: to ucp_init()"

File: src/ucp/core/ucp_context.c [hunk 2]
```diff
@@ -788,6 +796,11 @@ static ucs_status_t ucp_fill_config(ucp_context_h context,
                                               : UCP_MT_TYPE_SPINLOCK);
     context->config.ext = config->ctx;
 
+    if (context->config.ext.estimated_num_eps != UCS_CONFIG_MEMUNITS_AUTO) {
+        /* num_eps were set via the env variable. Override current value */
+        context->config.est_num_eps = context->config.ext.estimated_num_eps;
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: can add ucs_debug here

File: src/ucp/core/ucp_context.c [hunk 3]
```diff
@@ -170,6 +170,14 @@ static ucs_config_field_t ucp_config_table[] = {
    "Also the value has to be bigger than UCX_TM_THRESH to take an effect." ,
    ucs_offsetof(ucp_config_t, ctx.tm_max_bcopy), UCS_CONFIG_TYPE_MEMUNITS},
 
+  {"NUM_EPS", "auto",
+   "An optimization hint of how many endpoints would be created on this context.\n"
+   "Does not affect semantics, but only transport selection criteria and the\n"
+   "resulting performance.\n"
+   " If set to a value different from \"auto\" it will override the value passed\n"
+   "to the ucp_init()",
+   ucs_offsetof(ucp_config_t, ctx.estimated_num_eps), UCS_CONFIG_TYPE_MEMUNITS},
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: better add new type or extend UINT

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5455/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5475/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5478/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5479/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5481/ for details (Mellanox internal link).

**Pull request merged**